### PR TITLE
Keyboard shortcuts

### PR DESCRIPTION
**Pull request by sttz**
Hey, thanks for the great HTML5 player for Crunchyroll!

This PR adds a couple of additional keyboard shortcuts that I wish existed, mostly because I want to avoid using the mouse when watching on CR.

- I've added global keyboard shortcuts that don't need the player to have focus, avoiding to have to click on the player before being able to use the keyboard. For now that's just **f** to toggle fullscreen and **k** to toggle playback and they work just like the regular shortcuts. The global shortcuts aren't triggered when input fields have focus to avoid interfering with regular website usage.

- **shift-n** plays the next video. This wasn't possible without using the mouse. I had to refactor the functionality a bit so that the same logic could be used by the shortcut and the button.

- **s** skips ahead 85 seconds. This is based on the fact that most anime openings are exactly 90 seconds long and conveniently allows to skip them. It's five seconds shorter to add some slack and to avoid seeking too far.

I'm happy to revise the PR if you have any feedback and if you only want to cherry-pick some of the shortcuts, I can keep them in my fork.

P.S: I might be looking into integrating Trakt scrobbling. I see you have that already on your TODO-list in #3 but have you already started woking on it?

### CODE REVIEW

File: src/app/media/player/Player.tsx [hunk 1]
```diff
@@ -357,6 +357,16 @@ export class Player extends Component<IPlayerProps, {}> {
           api.unmute();
         }
         break;
+      // N
+      case 78:
+        if (e.shiftKey) {
+          api.playNextVideo();
+        }
```
Comment (YePpHa) [inline src/app/media/player/Player.tsx]: missing break

Comment (YePpHa) [inline src/app/media/player/Player.tsx]: missing break

File: src/app/media/player/chrome/NextVideoButton.tsx [hunk 2]
```diff
@@ -61,14 +61,8 @@ export class NextVideoButton extends Component<INextVideoButtonProps, {}> {
...
     const onClick = (e: Event) => {
-      const detail = this.props.api.getNextVideoDetail();
-      if (!detail) return;
-      const event = new NextVideoEvent(detail);
-      this.props.api.dispatchEvent(event);
-
-      if (event.defaultPrevented) {
-        e.preventDefault();
-      }
+      this.props.api.playNextVideo()
```
Comment (YePpHa) [inline src/app/media/player/chrome/NextVideoButton.tsx]: missing `;` at the end of the line

**Comment by YePpHa** [comment]
@sttz as far as tracking goes I need to figure out a generic way to implement them all, but before I want to start on that I want to implement a settings page for the player. The settings page will be where you will be able to set up the different tracking services.

**Comment by sttz** [comment]
Thanks! Added the missing `break;` in `Player.tsx` and the missing `;` in `NextVideoButton.tsx`.

File: src/app/media/player/Player.tsx [hunk 3]
```diff
@@ -369,6 +380,35 @@ export class Player extends Component<IPlayerProps, {}> {
     e.preventDefault();
   }
 
+  private _onDocumentKeyDown(e: BrowserEvent) {
+    const element = e.target as HTMLElement;
+    if (element && (element.tagName == 'INPUT' || element.tagName == 'SELECT' || element.tagName == 'TEXTAREA' || element.isContentEditable)) {
```
Comment (YePpHa) [inline src/app/media/player/Player.tsx]: Use triple equals as it's more consistent with the rest of the codebase.

File: src/app/media/player/Player.tsx [hunk 4]
```diff
@@ -369,6 +380,35 @@ export class Player extends Component<IPlayerProps, {}> {
...
+    }
+
+    const api = this.getApi();
+    switch (e.keyCode) {
+      // F
+      case 70:
+        api.toggleFullscreen();
+        break;
+      // K
+      case 75:
```
Comment (YePpHa) [inline src/app/media/player/Player.tsx]: The global shortcuts conflict with the player shortcuts. It should not trigger the global shortcuts if the focus is on the element or one of its elements.

**Comment by sttz** [comment]
Thanks for the catch! Changed to triple equals and added a check to prevent the global shortcuts from firing when the player has focus.

File: src/app/media/player/Player.tsx [hunk 5]
```diff
@@ -369,6 +380,38 @@ export class Player extends Component<IPlayerProps, {}> {
     e.preventDefault();
   }
 
+  private _onDocumentKeyDown(e: BrowserEvent) {
+    const element = e.target as HTMLElement;
+    if (element) {
+      if (element.tagName === 'INPUT' || element.tagName === 'SELECT' || element.tagName === 'TEXTAREA' || element.isContentEditable)
+        return;
+      if (this.base.contains(element))
```
Comment (YePpHa) [inline src/app/media/player/Player.tsx]: also check whether base is equal to element.

Comment (YePpHa) [reply src/app/media/player/Player.tsx]: Strange I remember that wasn't the case. Anyways, I will do a quick test when I get home to verify everything is working and then everything should be in order.

**Pull request merged**